### PR TITLE
[FLINK-28082][Connector/Pulsar] Add  end-to-end encryption support.

### PR DESCRIPTION
**Pull request by syhily**
This PR adds end-to-end encryption support in `flink-connector-pulsar`.

1. Add `CryptoKeyReader` setting for both `PulsarSourceBuilder` and `PulsarSinkBuilder`.
2. Add encrypt key setting for `PulsarSinkBuilder`.
3. Create a set of end-to-end encryption tests.
4. Add end to end documentation.

### CODE REVIEW

File: docs/content.zh/docs/connectors/datastream/pulsar.md [hunk 1]
```diff
@@ -804,113 +798,25 @@ Pulsar Sink 使用生产者 API 来发送消息。Pulsar 的 `ProducerConfigurat
...
+[Pulsar Schema Evolution][schema-evolution] 允许用户在一个 Flink 应用程序中使用的数据模型发生特定改变后（比如向基于 ARVO 的 POJO 类中增加或删除一个字段），仍能使用同一个 Flink 应用程序的代码。
 
-默认情况下，Pulsar 生产者每隔 60 秒才会刷新一次监控数据，然而 Pulsar Sink 每 500 毫秒就会从 Pulsar 生产者中获得最新的监控数据。因此 `numRecordsOut`、`numBytesOut`、`numAcksReceived` 以及 `numRecordsOutErrors` 4 个指标实际上每 60 秒才会刷新一次。
+可以在 Pulsar 集群内指定哪些类型的数据模型的改变是被允许的，详情请参阅 [Pulsar Schema Evolution][schema-evolution]。
 
-如果想要更高地刷新评率，可以通过如下方式来将 Pulsar 生产者的监控数据刷新频率调整至相应值（最低为1s）：
-{{< tabs "b65b9978-b1d6-4b0d-ade8-78098e0f23d1" >}}
+## 监控指标
+
+默认情况下，Pulsar client 每隔 60 秒才会刷新一次监控数据。如果想要更高地刷新评率，可以通过如下方式来将 Pulsar client 的监控数据刷新频率调整至相应值（最低为1s）：
```
Comment (tisonkun) [inline docs/content.zh/docs/connectors/datastream/pulsar.md:817]:
```suggestion
默认情况下，Pulsar client 每隔 60 秒才会刷新一次监控数据。如果想要更高地刷新频率，可以通过如下方式来将 Pulsar client 的监控数据刷新频率调整至相应值（最低为1s）：
```

File: flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/crypto/PulsarCrypto.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+     * @return It shouldn't be a null or empty instance if your have returned a key reader.
+     */
+    Set<String> encryptKeys();
+
+    /**
+     * {@link MessageCrypto} is used to define how to decrypt/encrypt the message. It's not required
+     * by default, because Pulsar will provide a default implementation by using the bouncy castle.
+     */
+    @Nullable
+    default MessageCrypto<MessageMetadata, MessageMetadata> messageCrypto() {
```
Comment (tisonkun) [inline flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/crypto/PulsarCrypto.java:72]: It seems we don't have test coverage for `messageCrypto != null`

Comment (syhily) [reply flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/crypto/PulsarCrypto.java:72]: I think this is a preserved method which is seldom used. I have a test for this by using the default `MessageCryptoBc` implementation.

Comment (tisonkun) [reply flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/crypto/PulsarCrypto.java:72]: Let's add one :)

Comment (syhily) [reply flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/crypto/PulsarCrypto.java:72]: Done, I add `MessageCrypto` in `ConsumeEncryptMessagesContext` and `PulsarEncryptSinkContext`. Can you check it?

**Review by tisonkun — APPROVED** [review]

**Comment by tisonkun** [comment]
Merging...

**Pull request merged**